### PR TITLE
[StructuredAuthenticationConfig] Add feature flag and wire up `--authentication-config` flag

### LINKED ISSUE #118833: Wire feature flag and to authn config struct
/sig auth
/assign @aramase 
/triage accepted
/milestone v1.28

### PR DESCRIPTION
**Pull request by aramase**
Implements the first phase of Structured Authentication Configuration
- Adds `StructuredAuthenticationConfiguration` feature gate
- kube-apiserver flag (`--authentication-config`) for authentication configuration file

fixes https://github.com/kubernetes/kubernetes/issues/118833

/kind feature
/sig auth
/triage accepted
/milestone v1.28
/priority important-soon

```release-note
kube-apiserver: adds --authentication-config flag for reading AuthenticationConfiguration files. --authentication-config flag is mutually exclusive with the existing --oidc-* flags.
```

```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3331-structured-authentication-configuration
```

### CODE REVIEW

**Comment by furkatgofurov7** [comment]
@aramase hey, Bug Triage Lead here 👋🏼

This still labelled as WIP though, JFYI, please be aware the code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about less than 2 weeks from now).

**Comment by aramase** [comment]
This PR includes builds on top of https://github.com/kubernetes/kubernetes/pull/118984 to add feature gate and wire kube-apiserver flag for reading authentication config.

**Comment by enj** [comment]
xref: reviewed this PR at https://github.com/aramase/kubernetes/pull/1

File: pkg/kubeapiserver/options/authentication.go [hunk 1]
```diff
@@ -584,3 +635,59 @@ func (o *BuiltInAuthenticationOptions) ApplyAuthorization(authorization *BuiltIn
 		o.Anonymous.Allow = false
 	}
 }
+
+func (o *BuiltInAuthenticationOptions) validateOIDCOptions() []error {
+	var allErrors []error
+
+	// Existing validation when jwt authenticator is configured with oidc-* flags
+	if len(o.AuthenticationConfigFile) == 0 {
+		if o.OIDC != nil && areOIDCFlagsConfigured() && (len(o.OIDC.IssuerURL) == 0 || len(o.OIDC.ClientID) == 0) {
```
Comment (deads2k) [inline pkg/kubeapiserver/options/authentication.go:644]: Why are the issuer URL and client ID check not required when the the AuthenticationConfigFile is specified?

File: pkg/kubeapiserver/options/authentication.go [hunk 2]
```diff
@@ -584,3 +635,59 @@ func (o *BuiltInAuthenticationOptions) ApplyAuthorization(authorization *BuiltIn
...
+	// read from file
+	data, err := os.ReadFile(configFilePath)
+	if err != nil {
+		return nil, err
+	}
+	if len(data) == 0 {
+		return nil, fmt.Errorf("empty config file %q", configFilePath)
+	}
+
+	install.Install(cfgScheme)
```
Comment (deads2k) [inline pkg/kubeapiserver/options/authentication.go:682]: this should only be done once. I can live with either an init block or a sync.Once

**Review by deads2k — COMMENTED** [review]
largely good, minor comments.

**Comment by aramase** [comment]
> largely good, minor comments.

thanks @deads2k! I've updated the PR to resolve the comments. PTAL!

File: pkg/kubeapiserver/options/authentication.go (hunk 1 above)
Comment (deads2k) [reply pkg/kubeapiserver/options/authentication.go]: found them in ValidateAuthenticationConfiguration .

**Pull request merged**